Spigot-API-Patches/0291-Add-hasMoved-to-PlayerMoveEvent.patch

```diff
@@ -0,0 +1,29 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: HexedHero <6012891+HexedHero@users.noreply.github.com>
+Date: Thu, 29 Apr 2021 10:31:44 +0100
+Subject: [PATCH] Add hasMoved to PlayerMoveEvent
+
+
+diff --git a/src/main/java/org/bukkit/event/player/PlayerMoveEvent.java b/src/main/java/org/bukkit/event/player/PlayerMoveEvent.java
+index 1a58734d919fae247eeb85dd785fd59990856505..a25556f52461b4e77b135c9f9fd30e3b6afdb08d 100644
+--- a/src/main/java/org/bukkit/event/player/PlayerMoveEvent.java
++++ b/src/main/java/org/bukkit/event/player/PlayerMoveEvent.java
+@@ -93,6 +93,18 @@ public class PlayerMoveEvent extends PlayerEvent implements Cancellable {
+         this.to = to;
+     }
+ 
++    // Paper start
++    /**
++     * Check if the player has actually moved somewhere and not just
++     * changed looking direction which also fires the event
++     *
++     * @return whether the player has moved or not
++     */
++    public boolean hasMoved() {
++        return from.getX() != to.getX() || from.getY() != to.getY() || from.getZ() != to.getZ();
```

Comment:
Could this be calculated when the event is initialised?

Reply:
I didn't to save calculations unless needed by the developer since the event is fired so often.

Reply:
Could be made a monotonic non null so that the result is saved once it's been called?

Reply:
Why would you save this information ? This is not realy heavy computation here. And if `to` is redefined by a listener the returned value would be non-sense.

Reply:
You could drop the cache on setTo call. Tho indeed, it might be not worth caching.